### ISSUE TITLE
ML HW

### ISSUE THREAD

**Issue by shuban-789** [shuban-789/Shuban-CSA#29]
Homework - Part 1

Similarities Between Java and Python Titanic ML Code

When comparing the Java Titanic ML implementation using Tablesaw/Weka and the Python implementation using Pandas/Scikit-learn, several similarities in structure and code flow can be observed:

1. Data Loading and Preprocessing

Java: Uses Tablesaw to load the CSV file and clean the data.

Python: Uses pandas.read_csv() followed by similar cleaning operations.

Common Steps:

Drop irrelevant columns (Name, Ticket, Cabin)

Convert categorical values (Sex, Embarked) to numeric representations

Fill missing values with medians

Normalize numerical features (Age, Fare, etc.)

2. Label Preparation

Java: Converts Survived into nominal labels ("Yes", "No") and identifies it as the class attribute.

Python: Keeps it as a binary target variable (0 or 1).

3. Model Training

Java: Uses Weka with classifiers like J48 and Logistic.

Python: Uses scikit-learn with models like DecisionTreeClassifier and LogisticRegression.

Commonality: Both fit two different models to the same dataset.

4. Evaluation

Java: Performs 10-fold cross-validation using weka.classifiers.Evaluation.

Python: Uses cross_val_score from sklearn.model_selection with cv=10.

5. Modular Structure

Java encapsulates parts like normalization and data conversion into helper methods.

Python separates preprocessing and modeling into reusable code cells or functions.

Summary:

Despite the difference in language and libraries, both versions follow:

Structured data preparation

Consistent feature engineering

Application of multiple classifiers

Cross-validation based model evaluation

Homework - Part 2

Graph Enhancements: Method Implementations

addNode(String node)

public void addNode(String node) {
    if (!graph.containsKey(node)) {
        graph.put(node, new HashMap<>());
    }
}

removeEdge(String from, String to)

public void removeEdge(String from, String to) {
    if (graph.containsKey(from)) {
        graph.get(from).remove(to);
    }
    if (graph.containsKey(to)) {
        graph.get(to).remove(from); // For undirected graphs
    }
}

Notes:

addNode ensures that the graph map includes a node with no edges.

removeEdge safely removes an edge between two nodes without deleting the nodes themselves.

These methods are essential for dynamically modifying a graph, useful in many real-world algorithms.
